File: src/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetElementById.java
package de.rub.bi.inf.openbimrl.functions.ifc;

import de.rub.bi.inf.openbimrl.NodeProxy;
import de.rub.bi.inf.openbimrl.engine.ifc.IIFCModel;
import de.rub.bi.inf.openbimrl.functions.AbstractFunction;

/**
 * Retrieves the IFC entity of a specific id from the model.
 *
 * @author Marcel Stepien
 */
public class GetElementById extends AbstractFunction {

    public GetElementById(NodeProxy nodeProxy) {
        super(nodeProxy);
    }

    @Override
    public void execute(IIFCModel ifcModel) {
        String guid = getInput(0);

        final var object = ifcModel.getObjectByID(guid);

        setResult(0, object);
    }

}

File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetElementById.kt
package de.rub.bi.inf.openbimrl.functions.ifc

import de.rub.bi.inf.nativelib.IfcPointer
import de.rub.bi.inf.nativelib.NativeEngine
import de.rub.bi.inf.openbimrl.NodeProxy
import de.rub.bi.inf.openbimrl.functions.NativeFunction
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionPort
import de.rub.bi.inf.openbimrl.functions.annotations.OpenBIMRLFunction

@OpenBIMRLFunction(
    name = "getElementById",
    description = "Retrieves an IFC element by its GlobalId (GUID).",
    inputs = [
        FunctionPort(0, "GUID", String::class),
    ],
    outputs = [
        FunctionPort(0, "IfcElement", IfcPointer::class),
    ],
)
class GetElementById(nodeProxy: NodeProxy) : NativeFunction(nodeProxy) {
    override fun executeNative() {
        val guid = getInput<String>(0) ?: return
        setResult(0, IfcPointer.fromHandle(NativeEngine.filterByGuid(guid)))
    }
}
